rep-2004.rst

```diff
@@ -0,0 +1,836 @@
+If there is a need for a particular package in a reasonable production scenario, then that package should be considered for this category as well.
+However, packages which we consider essential to getting a robot up and running quickly, but perhaps provide a generic solution to a problem should probably not start out as 'Level 1' due to the high effort required to get a package to 'Level 1' and maintain it there.
+
+For example, packages which provide intra-process communication, inter-process communication, generated message runtime code, node lifecycle, etc. should probably all be considered for 'Level 1'.
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+.. _Version Policy:
+
+1. **Version Policy:**
+
+   .. _1.i:
+
+   i. Must have a version policy (e.g. ``semver``)
+
+   .. _1.ii:
+
+   ii. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+
+   .. _1.iii:
+
+   iii. Must have a strictly declared public API
+
+   .. _1.iv:
+
+   iv. Must have a policy for API stability
+
+   .. _1.v:
+
+   v. Must have a policy for ABI stability
+
+   .. _1.vi:
+
+   vi. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+.. _Change Control Process:
+
+2. **Change Control Process:**
+
+   .. _2.i:
+
+   i. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+
+   .. _2.ii:
+
+   ii. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+
+   .. _2.iii:
+
+   iii. Must have peer review policy for all change requests (e.g. require one or more reviewers)
+
+   .. _2.iv:
+
+   iv. Must have Continuous Integration (CI) policy for all change requests
+
+   .. _2.v:
+
+   v. Must have documentation policy for all change requests
+
+.. _Documentation:
+
+3. **Documentation:**
+
+   .. _3.i:
+
+   i. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+
+   .. _3.ii:
+
+   ii. Must have documentation for each item in the public API (e.g. functions, classes, etc.)
+
+   .. _3.iii:
+
+   iii. Must have a declared license or set of licenses
+
+   .. _3.iv:
+
+   iv. Must have a copyright statement in each source file
```

Comment:
1. It is not necessary to have a copyright statement in each file for it to be copyrighted under US law
2. No project obeys this currently (CMakeLists.txt and setup.py are certainly source files and tend to never contain copyright)
3. The copyright info is usually incomplete. Why put it in a file header if you need to look at git history to see who *really* wrote a bit of code?
4. This prevents consumption to or from any code in the public domain.

If the intent is to attribute ownership of intellectual property, then it would be better to say that “the project must clearly attribute all authors and the intellectual property rights they maintain”.

Reply:
I've actually wondered about this as well.

Especially because of @clalancette's comment in https://github.com/ros2/ros2_documentation/issues/616#issuecomment-617420434:

> The git commits already contain this information

This REP seems to say something else.

I realise that @clalancette is "just one voice", but it seems even within one organisation there are different ideas about what is needed or not for the copyright management.

(besides that I'm really curious whether "but it's in the git commit history" holds up as an argument in a copyright claim or not, as files can be copied out of a git repository, leaving the git metadata behind, but that would be off-topic here)

Reply:
I agree this should be modified to be something like "the project must clearly attribute all authors and the intellectual property rights they maintain", as @rotu suggested, and then we can say that the way we happen to do that in ROS 2's core packages is by putting a license and copyright statement in each file that is also checked by our linter. What files are checked by the linter by default can be discussed there.

----

> > The git commits already contain this information 

> This REP seems to say something else.

How so? @clalancette's point, I think, was that if you want to know every contributor then the git history provides that. Most patches provided to our software get added under the existing copyright. If someone wants to add themselves to the copyright statements, that's always allowed, or if someone makes a new file they can have theirs exclusively. That's also allowed. I don't know of a project that requires every contributor to add themselves or their company to every file they touch, if that's the suggested alternative to using the git history for "who has contributed". I don't think this REP says something else either.

Reply:
I pushed 99c2865 to make this a little more compliant. The developer guide still says "all source files" and that it should be enforced by the "automated linter", so "what is a source file" can be defined/discussed there, and we get away from "what is good enough for copyright statements" and gets the REP back to "you need to handle this and say how you handle it" rather than "this is how you should handle it".